src/core/dappletContext.ts

```diff
@@ -20,20 +21,70 @@ export class DappletContext {
 
     constructor(config: ContextConfig = DEFAULT_CONFIG) {
         config = { ...DEFAULT_CONFIG, ...config };
-        this.dappletProviders = config.providers!;
+        this.dappletProviders = config.providers! || [];
         this.featureRegistry = new FeatureRegistry(config.features || []);
     }
 
     async processRequest(request: DappletRequest): Promise<DappletTxResult> {
+        // dapplet loading and validation
+        for (let i = 0; i < request.frames.length; i++) {
+            const dapplet = await this._loadDapplet(request.frames[i].dappletId);
+            this._validateDapplet(dapplet);
+            request.frames[i].dapplet = dapplet;
+        }
+
         const engine = new DappletEngine(request, this);
```

Comment:
I would not like to see a modified request as a parameter of `DappletEngine(...)`. 
You dont need a request, you need a dapplet and the parsed data for `DappletEngine`. 
```
const engine = new DappletEngine(dapplet:DappletTemplate, variablesdata:bytes[], this);
```

Reply:
It looks like you confuse DappletEngine and DappletFrameExecutor :)
The first receives frames to input.
The second receives dapplet and metadata to input as you want.

We can pass frames here